src/main/scala/org/constellation/checkpoint/CheckpointAcceptanceService.scala

```diff
@@ -314,14 +268,48 @@ class CheckpointAcceptanceService[F[_]: Concurrent: Timer](
     }
   }
 
+  def resolveMissingParents(
+    cb: CheckpointBlock,
+    peers: List[PeerApiClient],
+    depth: Int = 1
+  ): F[List[CheckpointCache]] = {
+
+    val checkError = if (depth >= maxDepth) {
+      Sync[F].raiseError[Unit](new Exception("Max depth reached when resolving data."))
+    } else Sync[F].unit
+
+    val soeHashes = cb.parentSOEHashes.toList
+    val existingSoeHashes = soeHashes
+      .traverse(checkpointParentService.soeService.lookup)
+      .map(_.flatten)
+      .map(_.map(_.signedObservationEdge.hash))
+    val missingSoeHashes = existingSoeHashes.map(existing => soeHashes.diff(existing))
+
+    val resolveCheckpoints = missingSoeHashes
+      .map(_.flatMap(_ => checkpointParentService.parentBaseHashesDirect(cb)))
+      .flatMap(_.traverse(hash => checkpointService.contains(hash).map(exist => (hash, exist))))
+      .map(_.filter(_._2))
+      .map(_.map(_._1))
```

Comment:
You can probably shorten it to single line:

```scala
.flatMap(_.filterA(checkpointService.contains))
```
`filterA[G]` takes effectful predicate function`String => G[Boolean]` and returns you already traversed  and filtered `G[Seq[String]]`